cmake/kconfig.cmake

```diff
@@ -49,12 +49,17 @@ foreach(kconfig_target ${kconfig_target_list})
     )
 endforeach()
 
+# Bring in extra configuration files dropped in by the user or anyone else;
+# make sure they are set at the end so we can override any other setting
```

Comment:
Thank you for contributing.

I saw that this was KBuild's behaviour. I didn't follow it because I assumed it was undesirable behaviour.

I need to think a bit about the implications.

Can you elaborate on this use-case:

"extra configuration files dropped in by the user or anyone else"

Isn't this use-case covered by CONF_FILE?

Reply:
Sorry, didn't see the commit message.

Can you elaborate on:

"even more to figure out how not to override CONF_FILE".

Reply:
We build very extensive parameterized test cases, and different independent parts take care of different parts.

For example:

- the main test framework makes sure that all code has BOOT_DELAY set to 1s, to make sure serial ports are ready to receive
- some other piece is deciding the stub code has to disable all boot banners and delays (which are forcibly enabled by IS_TESTCASE's overlay config), as otherwise it would conflict with the testcase's output -- the stub is supposed to be silent.
- other pieces are deciding (for example for network testcases) which runtime IP addresses are configured.

All these pieces work independently and in different parts of the flow to decide the configuration and then toss it all there; trying to unify with CONF_FILE becomes quite complicated as I might be overriding something somebody else has done before.

With this I can decide the right order by just setting the name, I don't have to create extra variable code to handle it and I can use the same code for pre-cmake and post-cmake. We are still running testcases for previous versions of Zephyr, so having to support two different systems is a no-no. It will just make things way more complicated.

An advertised and used feature shall be kept, not just dropped.

Reply:
I think I understand your use-case. Would you accept a compromise where
a dedicated directory

${APPLICATION_BINARY_DIR}/extra_configs/*

is used instead of the root directory?

The problem with wildcarding the root directory is that I have observed .conf files being accidentally included twice in the merge. The output looked like this:

> make[1]: Entering directory '/home/sebo/workspace_zephyr/zephyr/tests/subsys/fs/nffs_fs_api'
> Using /home/sebo/workspace_zephyr/zephyr/boards/x86/qemu_x86/qemu_x86_defconfig as base
> Merging /home/sebo/workspace_zephyr/zephyr/sanity-out/qemu_x86/tests/subsys/fs/nffs_fs_api/test_cache/overlay.conf
> Merging prj.conf
> Merging /home/sebo/workspace_zephyr/zephyr/sanity-out/qemu_x86/tests/subsys/fs/nffs_fs_api/test_cache/overlay.conf

If we have a dedicated folder for these untracked .conf files that the build system does not know about then I don't see any issues (except for having way too many mechanisms for specifying .conf files).

Reply:
that was my thought as well, have a dedicated directory rather than dump them into root and cause conflicts

Reply:
"An advertised and used feature shall be kept, not just dropped."
BTW, I strongly agree with you here, this is all backwards, we should
be convincing you about this change, not the other way around.

This feature was deliberately dropped with the intent to simplify an overly
complex system.

Reply:
why don't we 'sort -u' the list passed to the conf merger? -- not to mention that including twice (a bug anyway) should still be a idempotent operation

Reply:
if we have to resort to a subdirectory, I'd say yes in between teeth, because it'll still have a big impact -- I'd rather we fix the problem at the root

Reply:
/me finds [list(REMOVE_DUPLICATES)](https://cmake.org/cmake/help/v3.0/command/list.html?highlight=REMOVE_DUPLICATES); wouldn't that work?